compiler/rustc_trait_selection/src/traits/error_reporting/suggestions.rs

```diff
@@ -2668,15 +2668,49 @@ impl<'tcx> TypeErrCtxtExt<'tcx> for TypeErrCtxt<'_, 'tcx> {
                         if let DefKind::Trait = tcx.def_kind(item_def_id)
                             && !visible_item
                         {
-                            // FIXME(estebank): extend this to search for all the types that do
-                            // implement this trait and list them.
                             err.note(format!(
                                 "`{short_item_name}` is a \"sealed trait\", because to implement \
                                  it you also need to implement `{}`, which is not accessible; \
                                  this is usually done to force you to use one of the provided \
                                  types that already implement it",
                                 with_no_trimmed_paths!(tcx.def_path_str(def_id)),
                             ));
+                            let impls_of = tcx.trait_impls_of(def_id);
+                            let impls = impls_of
+                                .non_blanket_impls()
+                                .values()
+                                .flatten()
+                                .chain(impls_of.blanket_impls().iter())
+                                .collect::<Vec<_>>();
+                            match &impls[..] {
+                                [] => {}
+                                [only] => {
+                                    err.help(with_no_trimmed_paths!(format!(
+                                        "type `{}` implements the trait",
+                                        tcx.type_of(*only).instantiate_identity(),
+                                    )));
+                                }
+                                impls => {
+                                    let mut types = impls.iter()
+                                        .map(|t| with_no_trimmed_paths!(format!(
+                                            " - {}\n",
+                                            tcx.type_of(*t).instantiate_identity(),
+                                        )))
+                                        .collect::<Vec<_>>();
+                                    let post = if types.len() > 9 {
+                                        let len = types.len();
+                                        types.truncate(8);
```

Comment:
The truncate can be done unconditionally outside of `if`, then you also won't have to save `len`.

Reply:
We still need the `len` for correct grammar.